### PR TITLE
Add forward methods to quantizer that also computes commitment loss

### PR DESCRIPTION
**Pull request by hackyon**
# What does this PR do?

This PR adds the code to compute commitment loss for the quantizer. The loss is only computed in the newly added forward() methods.

This is a small part of the bigger #24295 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #24295 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

**Comment by hackyon** [comment]
Much of this was adapted from https://github.com/facebookresearch/encodec/blob/main/encodec/quantization/core_vq.py

For reference, a part of the conversation is in one of the specific changes as well:
https://github.com/huggingface/transformers/commit/4f697be0b62c4f3b0401ccbd00d1d46aac81906d

**Comment by hackyon** [comment]
@ArthurZucker 
Created a PR for this bit first, lemme know what you think. Thanks!

**Comment by sgugger** [comment]
cc @sanchit-gandhi

File: src/transformers/models/encodec/modeling_encodec.py [hunk 1]
```diff
@@ -372,6 +372,20 @@ def decode(self, embed_ind):
         quantize = nn.functional.embedding(embed_ind, self.embed)
         return quantize
 
+    def forward(self, hidden_states):
+        shape = hidden_states.shape
+
+        hidden_states = hidden_states.reshape((-1, shape[-1]))
+
+        embed_ind = self.quantize(hidden_states)
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:380]: Nice!

File: src/transformers/models/encodec/modeling_encodec.py [hunk 2]
```diff
@@ -392,6 +406,26 @@ def decode(self, embed_ind):
...
+
+        loss = torch.tensor([0.0], device=device, requires_grad=self.training)
+
+        if self.training:
+            # Pass the gradients straight through the quantization, by directly linking the gradients of the input
+            # embed_ind with the output quantize in the computation graph.
+            quantize = embed_ind + (quantize - embed_ind).detach()
+
+            commitment_loss = nn.functional.mse_loss(quantize.detach(), embed_ind)
+            loss = loss + commitment_loss
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:424]: Shall we add a config attribute for the commitment weight? Would only have this default to 1 if we are certain a user would not want to override this

What I would do in the init is set:
```python
self.commitment_weight = config.commitment_weight
```
Then in the forward:
```suggestion
            loss = loss + commitment_loss * self.commitment_weight
```

File: src/transformers/models/encodec/modeling_encodec.py [hunk 3]
```diff
@@ -392,6 +406,26 @@ def decode(self, embed_ind):
...
+
+        quantize = quantize.permute(0, 2, 1)
+        return loss, embed_ind, quantize
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:427]: Can we return an instance of `ModelOutput`, for example like we do in MBart:
https://github.com/huggingface/transformers/blob/134caef31a40eff778bb8dc7b917dac9d7e3177a/src/transformers/models/mbart/modeling_mbart.py#L1385

File: src/transformers/models/encodec/modeling_encodec.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/models/encodec/modeling_encodec.py:424]: Would actually be inclined to compute the loss as much as possible in the top level model (i.e. `EncodecModel`), rather than scattered across the sub-models. This should make the loss formulation easier to follow

File: src/transformers/models/encodec/modeling_encodec.py [hunk 4]
```diff
@@ -436,6 +470,27 @@ def decode(self, codes: torch.Tensor) -> torch.Tensor:
...
+        residual = embeddings
+        num_quantizers = self.get_num_quantizers_for_bandwidth(bandwidth)
+
+        quantized_out = torch.tensor(0.0, device=embeddings.device)
+        all_losses = []
+        all_indices = []
+
+        for layer in self.layers[:num_quantizers]:
+            loss, embed_ind, quantized = layer(residual)
+            # Note: The source from FB research did not detach the quantized here. There may have been a bug in the
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:483]: Interesting! My only fear with diverging from the original implementations is that we can't follow the hyper parameters given in the original implementation, e.g.
1. If a user has a working training set-up in the original codebase, they won't necessarily have a working one in Transformers
2. If we want to give sensible starting hyper parameters, we can't use the ones given in the paper -> we need to do our own HP search
3. If our implementation doesn't work, it's going to be hard asking the model authors how to fix this!

Maybe we can open an issue and check whether this is indeed a bug in the original?

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for picking this up so quickly @hackyon! Left some comments regarding the design, but you seem on the right tracks already 🤗

File: src/transformers/models/encodec/modeling_encodec.py (hunk 4 above)
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py:483]: Good point. I tend to agree that it would be better to follow the original implementation, even if there is a bug with it. The original issue in the FB research source is tracked here: https://github.com/facebookresearch/encodec/issues/25. I was able to reproduce the residual commitment losses being 0 in the latest code.

File: src/transformers/models/encodec/modeling_encodec.py (hunk 2 above)
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py:424]: Good point. I was looking through the code again, and I think it may be worth restructuring to actually compute the commitment loss during encode() rather than forward() for the EncodecEuclideanCodebook and EncodecVectorQuantization. 

This is a little off convention, but I believe it makes more sense since commitment loss depends only on encoding and quantization, but not decoding.

**Comment by hackyon** [comment]
Thanks for the review!

As I mentioned in https://github.com/huggingface/transformers/issues/24295#issuecomment-1614100206, I am busy in July so will be slow to iterate on this, but should have more time in August to keep iterating.

File: src/transformers/models/encodec/modeling_encodec.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/models/encodec/modeling_encodec.py:424]: Sounds good to me! Think it's nice to partition the losses this way (split between encoder and decoder), but wouldn't go any more partitioned than this to keep the loss computation easy to follow

**Comment by sanchit-gandhi** [comment]
Awesome @hackyon! I believe the Meta AI authors are planning on releasing more fine-tuning code throughout July, so we'll be in a good position to finish this PR on your return 🤗

**Comment by sanchit-gandhi** [comment]
Hey @hackyon - is this something you're till interested in working on? There was some really nice progress to kick-off this PR! Think it would make a nice contribution to have fine-tuning capabilities for the EnCodec model in `transformers`

**Comment by hackyon** [comment]
Thanks Sanchit. Sorry I've been a little busy as of late. I'll give it a good look again next week. I should have something out again before Sept 15th.

It would be nice to at least get the commitment loss computation in (though I might not have the free time to complete the discriminative loss part).

File: src/transformers/models/encodec/modeling_encodec.py (hunk 2 above)
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py]: I've updated the code to compute the commitment loss inside encode() rather than forward(), as it is dependent only on the encoding (during the quantization steps). 

I've also pulled it up to the top level, but this comes at the expense of more memory usage as the intermediate steps must now be stored and passed up to the top. 
1. What do you think about this? Do you think the extra clarify here is worth the memory overhead?
2. Should I just compute the loss inside _encode_frame(), so we only incur the memory overhead of the quantization_steps for one frame at a time?

Regarding the weight config, I think we can factor that in when we add the commitment_loss to the total loss (probably somewhere in the balancer step).

File: src/transformers/models/encodec/modeling_encodec.py (hunk 4 above)
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py]: FYI looks like this hasn't been fixed even in the updated audiocraft code:
https://github.com/facebookresearch/audiocraft/blob/a2b96756956846e194c9255d0cdadc2b47c93f1b/audiocraft/quantization/core_vq.py#L374C14-L374C14

**Comment by hackyon** [comment]
Please take a look and let me know what you think, thanks.

If the overall structure looks good, I'll write some tests and also come up with ways to test that the loss calculation we have here is equivalent to the one from FB audiocraft (feeding both models some inputs and checking that the output losses are equal).

**Comment by hackyon** [comment]
I also did some extra reading on the FB audiocraft code, and I think the ultimate training code would need to look something like this:
https://github.com/facebookresearch/audiocraft/blob/a2b96756956846e194c9255d0cdadc2b47c93f1b/audiocraft/solvers/compression.py#L83

With the balancer weights of the losses here:
https://github.com/facebookresearch/audiocraft/blob/a2b96756956846e194c9255d0cdadc2b47c93f1b/config/solver/compression/default.yaml#L14

The commitment_loss we are calculating in this change is part of the other_losses.


We'll need to add the balancer, the discriminator (adversarial loss), and some other spectrogram losses (aux_losses). There are also some code on the codebook updates that we might need to add as well (as they are updated using some kind of moving average process rather than through the general training step).

File: src/transformers/models/encodec/modeling_encodec.py (hunk 2 above)
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py]: I was thinking over this last night, and imho it's better to commit the commitment loss from within the quantizer and pass it back up (rather than computing it at the top level). The reasons are:

1. This mirrors the code in FB encodec/audiocraft, which will make it easier to maintain in the future
2. The quantization loss is a special case loss that depends only on the quantizer, and so it is somewhat natural for it to be there
3. Reduces unnecessary memory overhead

There will still be another area of code at the top level where we will add the commit loss into the other losses (adversarial, and spectrogram loss). 

Let me know what you think!

File: src/transformers/models/encodec/modeling_encodec.py [hunk 5]
```diff
@@ -59,6 +59,7 @@ class EncodecOutput(ModelOutput):
 
     audio_codes: torch.FloatTensor = None
     audio_values: torch.FloatTensor = None
+    commitment_loss: torch.FloatTensor = None
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:62]: This will need a docstring for the final PR :)

File: src/transformers/models/encodec/modeling_encodec.py [hunk 6]
```diff
@@ -411,21 +413,31 @@ def get_num_quantizers_for_bandwidth(self, bandwidth: Optional[float] = None) ->
...
+                quantized = residual + (quantized - residual).detach()
+                quantization_steps.append((residual, quantized.detach()))
+
+            # Note: There may be a bug here with the quantized results, but we do not fix it as it is present in the
+            # original FB code as well. For more context, see https://github.com/facebookresearch/encodec/issues/25.
             residual = residual - quantized
             all_indices.append(indices)
         out_indices = torch.stack(all_indices)
-        return out_indices
+        return out_indices, quantization_steps
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:440]: Can we convert `quantization_steps` to a tensor here? Like we do `all_indices`? Makes it easier to `torch.compile` a module!

File: src/transformers/models/encodec/modeling_encodec.py [hunk 7]
```diff
@@ -643,16 +656,24 @@ def encode(
...
             frame = input_values[:, :, offset : offset + chunk_length]
-            encoded_frame, scale = self._encode_frame(frame, bandwidth, mask)
+            encoded_frame, scale, quantization_steps = self._encode_frame(frame, bandwidth, mask)
             encoded_frames.append(encoded_frame)
+            quantization_steps_frames.append(quantization_steps)
             scales.append(scale)
 
         encoded_frames = torch.stack(encoded_frames)
 
+        commitment_loss = torch.tensor([0.0], device=input_values.device, requires_grad=self.training)
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:666]: Ideally `commitment_loss` is `None` if we're **not** training, and a torch tensor otherwise

File: src/transformers/models/encodec/modeling_encodec.py [hunk 8]
```diff
@@ -643,16 +656,24 @@ def encode(
...
+        if self.training:
+            for quantization_steps in quantization_steps_frames:
+                for residual, quantize in quantization_steps:
+                    loss = nn.functional.mse_loss(quantize.permute(0, 2, 1), residual.permute(0, 2, 1))
+                    commitment_loss = commitment_loss + loss
+
         if not return_dict:
             return (encoded_frames, scales)
```
Comment (sanchit-gandhi) [inline src/transformers/models/encodec/modeling_encodec.py:674]: `return_dict` output needs to match that of the data class output
```suggestion
            return (encoded_frames, scales, commitment_loss)
```

**Review by sanchit-gandhi — COMMENTED** [review]
Thanks for iterating @hackyon! It's great to have you back working on this PR 🤗 Code for computing the loss is more or less good with me! Just a bit worried about a potential breaking change on the output signature (see below)

File: src/transformers/models/encodec/modeling_encodec.py (hunk 2 above)
Comment (sanchit-gandhi) [reply src/transformers/models/encodec/modeling_encodec.py]: Good with me, particularly for cleaner code and bypassing unnecessary complexity with passing the decoded latents around from the quantiser to the main encode method

File: src/transformers/models/encodec/modeling_encodec.py [hunk 9]
```diff
@@ -801,11 +822,12 @@ def forward(
         if audio_scales is not None and audio_codes is None:
             raise ValueError("You specified `audio_scales` but did not specify the `audio_codes`")
 
+        commitment_loss = None
         if audio_scales is None and audio_codes is None:
-            audio_codes, audio_scales = self.encode(input_values, padding_mask, bandwidth, False)
+            audio_codes, audio_scales, commitment_loss = self.encode(input_values, padding_mask, bandwidth, False)
```
[the remark this answers is not in the document]
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py:827]: I believe we can just make the compute_loss type Optional in this case, which should allow for backwards compatibility.

Looking at some of the other models, many of them use a generic Tuple[Tensor] type for when return_dict = True. I think we can make the new third Optional parameter take on this type too, to future-proof the return type.

What do you think?

[the remark this answers is not in the document]
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py:827]: Technically, it would still break any existing unpacking unless they caller specifically adds in the indices to unpack (ie. audio_codes, audio_scales = output_tuple[:2]). 

With that said, I went through some of the other models, and that seems to be the convention.

Out of curiosity, is the purpose of return_dict just for backwards compatibility? Would it be possible to just deprecate the return_dict? I believe JIT already supports dataclasses return types.

[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply src/transformers/models/encodec/modeling_encodec.py:827]: It's this breaking of the unpacking that I'm a bit worried about - this would break the model for existing users. Most people won't be looking at the source code, so won't be aware of the change.

Maybe it's easier to add the loss computation at the top level model? Then we can add a `loss` entry to the `EncodecOutput` and won't break the encode method?

[the remark this answers is not in the document]
Comment (hackyon) [reply src/transformers/models/encodec/modeling_encodec.py:827]: I see. I will give that a try, and see how that works out.

[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/encodec/modeling_encodec.py:827]: Yep adding at the top level is probably better! I don't have other ideas on how to make it compatible 😞

**Comment by sanchit-gandhi** [comment]
Hey @hackyon - if in the future you find the time to continue on this PR and want to jump back into it, myself and Arthur would be more than happy to assist you with the final bits of the integration! You've done a nice job at setting up all the losses, so it's just a case of finalising the API and ensuring backwards compatibility!

**Comment by hackyon** [comment]
Thanks @sanchit-gandhi! I'll revisit this again when I find the time.
